solutions/dynamic_programming/unique_paths.py

```diff
@@ -0,0 +1,111 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Problem
+    A robot is located at the top-left corner of a 'M x N' grid (Marked 'Start'
+    in the diagram below).
+
+    The robot can only move either down or right at any point in time. The
+    robot is trying to reach the bottom-right corner of the grid (Marked
+    'Finish' in the diagram below).
+
+    How many possible unique paths are there?
+
+    Diagram
+
+        _____________________________
+        | R |   |   |   |   |   |   |
+        _____________________________
+        |   |   |   |   |   |   |   |
+        _____________________________
+        |   |   |   |   |   |   |   |
+        -----------------------------
+                                  ^
+                               'Finish'
+
+    Above is a 7 x 3 gridd. How many possible unique paths are there?
+
+    Note: M and N will be at most 100.
+
+Example 1
+
+    Input
+
+        M = 3, N = 2
+
+    Output
+
+        3
+
+    Explanation
+
+        From the top-left corner, there are a total of 3 ways to reach the
+        bottom-right corner:
+            1. Right -> Right -> Down
+            2. Right -> Down -> Right
+            3. Down -> Right -> Right
+
+Example 2
+
+    Input
+
+        M = 7, N = 3
+
+    Output
+
+        28
+"""
+from typing import Tuple, Dict, Union
+
+Row_index = int
+Column_index = int
+Cell = Tuple[Row_index, Column_index]
+VisitedPaths = Union[Dict[Cell, int], None]
+NumberOfPaths = int
+
+
+ROW = 0
+COLUMN = 1
+
+
+def _count_paths(
```

Comment:
my preference for parameter order is:

input params, state params, output params

in this case, the order i would choose is:

`total_rows, total_columns, cur_row, cur_col, visited`